### PR TITLE
[s] experimental automated game master

### PR DESCRIPTION
**Pull request by xxalpha**
experimental automated game master

### CODE REVIEW

**Comment by Bawhoppen** [comment]
:game_die:

**Comment by KazeEspada** [comment]
Can it choose space?

**Comment by xxalpha** [comment]
who knows

it is experimental

File: code/controllers/subsystem/gamemaster.dm [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+	gamemaster_announce("The engineering department has started a singularity at [T.loc].")
+	new /obj/singularity(T)
+
+/datum/subsystem/gamemaster/proc/el_dorado()
+	var/area/A = safepick(z1_areas)
+	if(A)
+		for(var/turf/T in A)
+			if(T.density)
+				T.ChangeTurf(/turf/simulated/wall/mineral/gold)
```
Comment (Supermichael777) [inline code/controllers/subsystem/gamemaster.dm]: wont this hit doors?

File: code/controllers/subsystem/gamemaster.dm [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+				shake_camera(M, 2, 1) // turn it down a bit come on
+			else
+				shake_camera(M, 7, 1)
+				if(istype(M))
+					M.Weaken(3)
+
+/datum/subsystem/gamemaster/proc/change_dir()
+	var/direction = pick(SOUTH, EAST, WEST)
+	for(var/client/C in clients)
+		if(!(C in admins))
```
Comment (Supermichael777) [inline code/controllers/subsystem/gamemaster.dm]: favoritism

File: code/controllers/subsystem/gamemaster.dm [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+			C.dir = direction
+	gamemaster_announce("This will take some time for you to get used to...")
+	spawn(3600)
+		for(var/client/C in clients)
+			C.dir = NORTH
```
Comment (Iamgoofball) [inline code/controllers/subsystem/gamemaster.dm]: @yota

File: code/controllers/subsystem/gamemaster.dm [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+			floor.ChangeTurf(/turf/simulated/floor/fakespace)
+			CHECK_TICK
+	gamemaster_announce("I upgraded all the station flooring so you can marvel at the beauty of space.")
+
+/datum/subsystem/gamemaster/proc/make_antag()
+	var/mob/M = pick(player_list)
+	var/antag_type = pick("Traitor", "Changeling", "Wizard", "Revolutionary", "Cultist")
+	switch(antag_type)
+		if("Traitor")
+			M.mind.make_Traitor()
```
Comment (Supermichael777) [inline code/controllers/subsystem/gamemaster.dm]: do these equip people?

File: code/controllers/subsystem/gamemaster.dm [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+/datum/subsystem/gamemaster/proc/dwarves_n_giants()
+	var/list/dwarves = player_list.Copy()
+	var/list/giants = list()
+	var/half = abs(dwarves.len / 2)
+	while(half > 0)
+		half--
+		giants += pick_n_take(dwarves)
+
+	for(var/V in dwarves)
```
Comment (Supermichael777) [inline code/controllers/subsystem/gamemaster.dm]: give them the dwarfism mutation.

File: code/controllers/subsystem/gamemaster.dm [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+		var/mob/M = V
+		M.resize = 0.8
+		M.update_transform()
+		M.pass_flags |= PASSTABLE
+	for(var/V in giants)
+		var/mob/M = V
+		M.resize = 1.5
+		M.update_transform()
+		M.pass_flags |= PASSTABLE
+	gamemaster_announce("At least none of you are elves.")
```
Comment (Supermichael777) [inline code/controllers/subsystem/gamemaster.dm]: best anouncment of the lot

File: code/controllers/subsystem/gamemaster.dm [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+	for(var/mob/M in player_list)
+		if(!istype(M,/mob/new_player) && !M.ear_deaf)
+			M << announcement
+			if(M.client.prefs.toggles & SOUND_ANNOUNCEMENTS)
+				continue
+			if(sound)
+				M << sound('sound/misc/announce_dig.ogg')
+
+/datum/subsystem/gamemaster/proc/bwoink()
+	playsound_global('sound/effects/adminhelp.ogg', 0, channel=1, volume=100)
```
Comment (Supermichael777) [inline code/controllers/subsystem/gamemaster.dm]: 100 is realy loud

**Comment by xxalpha** [comment]
stop commenting on this code, you are not allowed to look at experimental code

**Comment by PKPenguin321** [comment]
so a robot admin?
>inb4 it gains sentience and bans the rest of the admin team

**Comment by Mark9013100** [comment]
This is how SKYNET begins.

File: code/controllers/subsystem/gamemaster.dm (hunk 1 above)
Comment (Dorsisdwarf) [inline code/controllers/subsystem/gamemaster.dm]: Doors aren't turfs

**Comment by Jazaen** [comment]
Just make sure it doesn't learn from in game and OOC messages, or we will be screwed.

File: code/controllers/subsystem/gamemaster.dm (hunk 3 above)
Comment (YotaXP) [inline code/controllers/subsystem/gamemaster.dm]: Darnit.  Why has everyone been wanting to screw with client.dir lately?  Here's your filthy icon smoothing fix. https://gist.github.com/YotaXP/4b996eb96f4a39b0ea0a

*Does not fix bar sign.
*Does not fix "Space Station 13" logo on floor.
*Does not fix derpy maps where some turfs have dir=0. (I'm looking at you, tiles in front of medbay on Box)

Comment (xxalpha) [inline code/controllers/subsystem/gamemaster.dm]: no, this is experimental

**Comment by AlexanderUlanH** [comment]
Is it just me or is this functionally very similar to events?

**Comment by xxalpha** [comment]
There difference is that these are bad events.

File: code/controllers/subsystem/gamemaster.dm [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+			M.mind.make_Traitor()
+		if("Changeling")
+			M.mind.make_Changling()
+		if("Wizard")
+			M.mind.make_Wizard()
+		if("Revolutionary")
+			M.mind.make_Rev()
+		if("Cultist")
+			M.mind.make_Cultist()
+	gamemaster_announce("To spice things up, I have turned [M] into a [antag_type].")
```
Comment (Shadowlight213) [inline code/controllers/subsystem/gamemaster.dm]: Oh god. Announcing who the antag is. Insta lynch,

File: code/controllers/subsystem/gamemaster.dm [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+			if("mecha_warfare")
+				mecha_warfare()
+			if("clumsy")
+				clumsy()
+			if("dwarves_n_giants")
+				dwarves_n_giants()
+			if("investigate")
+				investigate()
+			if("shuttle")
+				shuttle()
```
Comment (duncathan) [inline code/controllers/subsystem/gamemaster.dm]: you could replace this switch case with call()() if you made the list have the actual procs

Comment (duncathan) [inline code/controllers/subsystem/gamemaster.dm]: why am I reviewing this

Comment (xxalpha) [inline code/controllers/subsystem/gamemaster.dm]: this is experimental

Comment (Iamgoofball) [inline code/controllers/subsystem/gamemaster.dm]: isnt call() heresy tho

**Comment by duncathan** [comment]
I guess it's experimental

**Comment by xxalpha** [comment]
I've gathered all the data possible to improve on this experimental feature.

**Comment by Jazaen** [comment]
No. Let it rot in hell, lest another April Fools is either ruined or runs on normal codebase

**Comment by xxalpha** [comment]
too late

**Comment by Iamgoofball** [comment]
Last time we did this all our april fools shit got thrown in the trash over the fake admin, are we sure we want to do this?

**Pull request merged**